### PR TITLE
WIP Add post-PPD rewrite

### PR DESCRIPTION
**Pull request by kasiafi**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/IterativeOptimizer.java [hunk 1]
```diff
@@ -182,15 +188,25 @@ private boolean exploreNode(int group, Context context)
...
         PlanNode expression = context.memo.getNode(group);
-        for (PlanNode child : expression.getSources()) {
+        for (int i = 0; i < expression.getSources().size(); i++) {
+            PlanNode child = expression.getSources().get(i);
             checkState(child instanceof GroupReference, "Expected child to be a group reference. Found: " + child.getClass().getName());
 
-            if (exploreGroup(((GroupReference) child).getGroupId(), context)) {
+            Context childContext;
+            if (i == 0) {
+                // pass the context of Delete to the left branch of plan
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/IterativeOptimizer.java:202]: Maybe more elegant way would be to add `boolean forDelete` to `TableScanNode`. Then we could simply use `PlanNodeSearcher` within a rule to search for `TableScans` with such flag.

In current approach `delete` logic (context goes to left) becomes part of `IterativeOptimizers` itself

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDereferencePushDown.java [hunk 2]
```diff
@@ -38,6 +40,11 @@
 public class TestDereferencePushDown
         extends BasePlanTest
 {
+   /* public TestDereferencePushDown()
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDereferencePushDown.java:43]: remove?

File: presto-main/src/test/java/io/prestosql/sql/planner/TestQuantifiedComparison.java [hunk 3]
```diff
@@ -35,7 +37,12 @@
     public void testQuantifiedComparisonEqualsAny()
     {
         String query = "SELECT orderkey, custkey FROM orders WHERE orderkey = ANY (VALUES ROW(CAST(5 as BIGINT)), ROW(CAST(3 as BIGINT)))";
-        assertPlan(query, anyTree(
+        assertPlan(
+                query,
+                Session.builder(getQueryRunner().getDefaultSession())
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestQuantifiedComparison.java:42]: Do we care that it's not rewritten? In other tests it seem to make sense, but in this one not neccecerly.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestLogicalPlanner.java [hunk 4]
```diff
@@ -1555,4 +1557,11 @@ private Session automaticJoinDistribution()
                 .setSystemProperty(JOIN_DISTRIBUTION_TYPE, JoinDistributionType.AUTOMATIC.name())
                 .build();
     }
+
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestLogicalPlanner.java:1560]: add a test case for semi join rewrite here.

**Comment by kasiafi** [comment]
Applied comments.
@sopel39, according to your suggestion, delete context is now property of TableScanNode instead of being passed from DeleteNode.

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java [hunk 5]
```diff
@@ -201,7 +202,8 @@ protected RelationPlan visitTable(Table node, Void context)
             }
 
             List<Symbol> outputSymbols = outputSymbolsBuilder.build();
-            PlanNode root = TableScanNode.newInstance(idAllocator.getNextId(), handle, outputSymbols, columns.build());
+            boolean isDeleteTarget = analysis.isDeleteTarget(createQualifiedObjectName(session, node, node.getName()));
+            PlanNode root = TableScanNode.newInstance(idAllocator.getNextId(), handle, outputSymbols, columns.build(), isDeleteTarget);
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java:206]: is there any way to test it?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                context.getIdAllocator().getNextId(),
+                semiJoin.getFilteringSource(),
+                ImmutableMap.of(),
+                singleGroupingSet(ImmutableList.of(semiJoin.getFilteringSourceJoinSymbol())),
+                ImmutableList.of(),
+                SINGLE,
+                Optional.empty(),
+                Optional.empty());
+
+        Expression joinFilter = new ComparisonExpression(EQUAL, semiJoin.getSourceJoinSymbol().toSymbolReference(), semiJoin.getFilteringSourceJoinSymbol().toSymbolReference());
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java:129]: nit: why not to create equi condition immediatelly?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+        JoinNode innerJoin = new JoinNode(
+                semiJoin.getId(),
+                INNER,
+                semiJoin.getSource(),
+                filteringSourceDistinct,
+                ImmutableList.of(),
+                semiJoin.getSource().getOutputSymbols(),
+                ImmutableList.of(),
+                Optional.of(joinFilter),
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java:139]: why not to put `filteredPredicate` immediatelly?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+        ProjectNode project = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                innerJoin,
+                Assignments.builder()
+                        .putIdentities(innerJoin.getOutputSymbols())
+                        .put(semiJoinSymbol, TRUE_LITERAL)
+                        .build());
+
+        if (filteredPredicate.equals(TRUE_LITERAL)) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java:157]: You probably could skip generating of `filteredPredicate` and just put entire predicate in `io.prestosql.sql.planner.plan.JoinNode#filter` (assuming true for `isSemiJoinSymbol`). That would simplify rule.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestQuantifiedComparison.java [hunk 9]
```diff
@@ -36,11 +39,11 @@ public void testQuantifiedComparisonEqualsAny()
     {
         String query = "SELECT orderkey, custkey FROM orders WHERE orderkey = ANY (VALUES ROW(CAST(5 as BIGINT)), ROW(CAST(3 as BIGINT)))";
         assertPlan(query, anyTree(
-                    filter("S",
-                            project(
-                                    semiJoin("X", "Y", "S",
-                                            anyTree(tableScan("orders", ImmutableMap.of("X", "orderkey"))),
-                                            anyTree(values(ImmutableMap.of("Y", 0))))))));
+                join(
+                        JoinNode.Type.INNER,
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestQuantifiedComparison.java:43]: static import `INNER`

**Review by sopel39 — APPROVED** [review]
lgtm % small comments

File: presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java (hunk 5 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/RelationPlanner.java:206]: I didn't find a way. DELETE queries are rejected in `TestLogicalPlanner`

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java (hunk 7 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java]: `filteredPredicate` might depend on the semi join symbol.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java (hunk 8 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java]: done

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                                    p.values(b),
+                                    a,
+                                    b,
+                                    aInB,
+                                    Optional.empty(),
+                                    Optional.empty(),
+                                    Optional.empty()));
+                })
+                .matches(project(
+                        ImmutableMap.of("a", PlanMatchPattern.expression("a"), "a_in_b", PlanMatchPattern.expression("true")),
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java:59]: static import `expression`

**Review by sopel39 — APPROVED** [review]

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java (hunk 10 above)
Comment (kasiafi) [reply presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java:59]: Can't import it due to collision with `PlanBuilder.expression()`

**Comment by sopel39** [comment]
[Benchmarks comparison-semi_part.pdf](https://github.com/prestosql/presto/files/5494168/Benchmarks.comparison-semi_part.pdf)
[Benchmarks comparison-semi_unpart.pdf](https://github.com/prestosql/presto/files/5494169/Benchmarks.comparison-semi_unpart.pdf)

significant improvements for: `tpcds/q83, tpcds/q95, tpcds/q58`, no regressions

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.singleGroupingSet;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.values;
+import static io.prestosql.sql.planner.iterative.rule.test.PlanBuilder.expression;
+import static io.prestosql.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.INNER;
+
+public class TestTransformFilteringSemiJoinToInnerJoin
+        extends BaseRuleTest
+{
+    @Test
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java:38]: Could we have a test that the rule does not fire with `forDelete` TS?
Are there tests for DELETE query with IN (semi-join)?

Comment (kasiafi) [reply presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestTransformFilteringSemiJoinToInnerJoin.java:38]: There's `testDelete()` in `AbstractTestDistributedQueries`. It covers the case of SemiJoin under DeleteNode. It runs for Raptor.

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 12]
```diff
@@ -482,6 +483,17 @@ public CorrelatedJoinNode correlatedJoin(List<Symbol> correlation, PlanNode inpu
         return new CorrelatedJoinNode(idAllocator.getNextId(), input, subquery, correlation, type, filter, originSubquery);
     }
 
+    public TableScanNode tableScan(List<Symbol> symbols, boolean forDelete)
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java:486]: nit: you could probably extend some exiting method and add an override, but it's also fine

Comment (kasiafi) [reply presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/PlanBuilder.java:486]: I decided to add a separate method for now. I think we could reorganize the `tableScan()` methods later when `forDelete` property is more used and tested.

**Comment by kbajda** [comment]
@sopel39 the benchmarks show usual regressions in TPC-H (unpart) for Q6 and Q20. 
There is also a small regression for Q67 (partitioned). Any insights there?

**Comment by sopel39** [comment]
merged, thanks!

**Comment by sjx782392329** [comment]
What has this pr improved? I can't understand without relevant description
`Improve performance of queries with uncorrelated IN clauses`

**Comment by kasiafi** [comment]
@sjx782392329 uncorrelated IN clauses are planned as `SemiJoinNode`. This PR adds an optimizer rule which captures `SemiJoinNode` and transforms it into join. A `JoinNode`, unlike `SemiJoinNode`, can be further optimized (CBO) and eventually produce a better plan. 
Here are benchmarks which show that this change significantly improved performance in certain cases: https://github.com/prestosql/presto/pull/5582#issuecomment-722333004

**Comment by guiyanakuang** [comment]
@kasiafi, I observe that the uncorrelated `not in` clauses are also planned for SemiJoinNode. This pr does not transform such a `SemiJoinNode` into a `JoinNode`.  I wonder if eventually transforming the uncorrelated `not in` clause into a `JoinNode(type = left)` would provide the same benefit?

**Pull request merged**